worker/upgradeseries/worker.go

```diff
@@ -0,0 +1,334 @@
+	// a *specific* unit, so we need to ensure they are *all* stopped and only
+	// perform the status updates afterwards. We can't afford to be in a
+	// partial state.
+	for unit, serviceName := range unitServices {
+		svc, err := w.service.DiscoverService(serviceName)
+
+		running, err := svc.Running()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if !running {
+			continue
+		}
+
+		if err := svc.Stop(); err != nil {
+			return errors.Annotatef(err, "stopping %q unit agent for series upgrade", unit)
+		}
+	}
+
+	return errors.Trace(w.SetMachineStatus(model.UpgradeSeriesPrepareMachine))
+}
+
+// transitionPrepareMachine rewrites service unit files for unit agents running
+// on this machine so that they are compatible with the init system of the
+// series upgrade target
+func (w *upgradeSeriesWorker) transitionPrepareComplete(statusCount int) error {
+	w.logger.Logf(loggo.INFO, "preparing service units for series upgrade")
+
+	_, err := w.unitAgentServices(statusCount)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// TODO (manadart 2018-08-09): Unit file wrangling to come.
+	// For now we just update the machine status to progress the workflow.
+
+	return errors.Trace(w.SetMachineStatus(model.UpgradeSeriesPrepareComplete))
+}
+
+// transitionUnitsStarted iterates over units managed by this machine.
+// It changes each unit state start the completion of the series upgrade,
+// then starts the unit's agent service.
+func (w *upgradeSeriesWorker) transitionUnitsStarted(statusCount int) error {
+	w.logger.Logf(loggo.INFO, "starting units after series upgrade")
+
+	unitServices, err := w.unitAgentServices(statusCount)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for unit, serviceName := range unitServices {
+		svc, err := w.service.DiscoverService(serviceName)
+
+		// TODO (manadart 2018-08-09): Add a method to the API to do this for
+		// all units at once.
+		if err := w.setUnitStatus(unit, model.CompleteStatus, model.UnitStarted); err != nil {
+			return errors.Trace(err)
+		}
+
+		running, err := svc.Running()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if running {
+			continue
+		}
+
+		if err := svc.Start(); err != nil {
+			return errors.Annotatef(err, "starting %q unit agent after series upgrade", unit)
+		}
+	}
+
+	return nil
+}
+
+// unitAgentServices filters the services running on the local machine to those
+// that are for unit agents. If the number of services returned differs from
+// input count, then a error is returned.
+func (w *upgradeSeriesWorker) unitAgentServices(statusCount int) (map[string]string, error) {
+	services, err := w.service.ListServices()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	unitServices := service.FindUnitServiceNames(services)
+	if len(unitServices) != statusCount {
+		return nil, errors.Errorf("mismatched counts; upgrade-series statuses: %d, detected services: %d",
+			statusCount, len(unitServices))
```

Comment:
is it worth logging/including the actual list of things that you discovered?

Is this only necessary because we've lost the names of things we wanted to look at and only have a count?

Reply:
Here we want to find the init system services (so we can start/stop them) and see if they match the amount of units we thing we have - Its a sanity check.

Reply:
We can log the discovered services.